crates/primitives/src/utils/mod.rs

```diff
@@ -0,0 +1,122 @@
+//! Common Ethereum utilities.
+
+use crate::B256;
+use alloc::vec::Vec;
+use core::mem::MaybeUninit;
+
+mod units;
+pub use units::{
+    format_ether, format_units, parse_ether, parse_units, ParseUnits, Unit, UnitsError,
+};
+
+#[doc(hidden)]
+#[deprecated(since = "0.5.0", note = "use `Unit::ETHER.wei()` instead")]
+pub const WEI_IN_ETHER: crate::U256 = Unit::ETHER.wei_const();
+
+#[doc(hidden)]
+#[deprecated(since = "0.5.0", note = "use `Unit` instead")]
+pub type Units = Unit;
+
+/// The prefix used for hashing messages according to EIP-191.
+pub const EIP191_PREFIX: &str = "\x19Ethereum Signed Message:\n";
+
+/// Hash a message according to [EIP-191] (version `0x01`).
+///
+/// The final message is a UTF-8 string, encoded as follows:
+/// `"\x19Ethereum Signed Message:\n" + message.length + message`
+///
+/// This message is then hashed using [Keccak-256](keccak256).
+///
+/// [EIP-191]: https://eips.ethereum.org/EIPS/eip-191
+pub fn eip191_hash_message<T: AsRef<[u8]>>(message: T) -> B256 {
+    keccak256(eip191_message(message))
+}
+
+/// Constructs a message according to [EIP-191] (version `0x01`).
+///
+/// The final message is a UTF-8 string, encoded as follows:
+/// `"\x19Ethereum Signed Message:\n" + message.length + message`
+///
+/// [EIP-191]: https://eips.ethereum.org/EIPS/eip-191
+pub fn eip191_message<T: AsRef<[u8]>>(message: T) -> Vec<u8> {
+    fn eip191_message(message: &[u8]) -> Vec<u8> {
+        let len = message.len();
+        let mut len_string_buffer = itoa::Buffer::new();
+        let len_string = len_string_buffer.format(len);
+
+        let mut eth_message = Vec::with_capacity(EIP191_PREFIX.len() + len_string.len() + len);
+        eth_message.extend_from_slice(EIP191_PREFIX.as_bytes());
+        eth_message.extend_from_slice(len_string.as_bytes());
+        eth_message.extend_from_slice(message);
+        eth_message
+    }
+
+    eip191_message(message.as_ref())
+}
+
+/// Simple interface to the [`Keccak-256`] hash function.
+///
+/// [`Keccak-256`]: https://en.wikipedia.org/wiki/SHA-3
+pub fn keccak256<T: AsRef<[u8]>>(bytes: T) -> B256 {
+    fn keccak256(bytes: &[u8]) -> B256 {
+        let mut output = MaybeUninit::<B256>::uninit();
+
+        cfg_if::cfg_if! {
+            if #[cfg(all(feature = "native-keccak", not(feature = "tiny-keccak")))] {
```

Comment:
do we want to throw a compile error if both features are used?

Reply:
No, we already went through this when this was added